### PR TITLE
[WIP] marshal mongodb bson into relaxed json

### PR DESCRIPTION
**Pull request by mukundansundar**
Signed-off-by: Mukundan Sundararajan <msundar.ms@outlook.com>

# Description

Change to marshal to Ext JSON with Relaxed format instead of Canonical Format. 

Without the change to `canonical` representation as false, the added conformance test scenario fails 
```
state.go:257: Checking value presence for b92a399b3889432782b78babd1169ab8-struct-with-int
    state.go:675: 
        	Error Trace:	state.go:675
        	            				state.go:266
        	Error:      	unmarshal error
        	Test:       	TestStateConformance/mongodb/get
        	Messages:   	error: %!w(*json.UnmarshalTypeError=&{object 0x10721fe20 12 IntValueType message}), json: {"message":{"$numberInt":"42"}}
```

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1582 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Review by halspang — COMMENTED** [review]
Will this change in encoding get to any user's code? Will they have to handle it differently? Or will this all go unnoticed by them?

File: state/mongodb/mongodb.go [hunk 1]
```diff
@@ -225,7 +225,14 @@ func (m *MongoDB) Get(req *state.GetRequest) (*state.GetResponse, error) {
 	case string:
 		data = []byte(obj)
 	case primitive.D:
-		if data, err = bson.MarshalExtJSON(obj, true, true); err != nil {
+		// Setting canonical to `false`.
```
Comment (artursouza) [inline state/mongodb/mongodb.go:228]: This seems to be a breaking change. If this component is Alpha, it is OK.

Comment (mukundansundar) [reply state/mongodb/mongodb.go:228]: It is not alpha ... It is stable component since v1.0 ...

**Comment by mukundansundar** [comment]
> Will this change in encoding get to any user's code? Will they have to handle it differently? Or will this all go unnoticed by them?

In general, when an ID field with decimal type is stored, the user expects the ID field to be returned as a decimal not as an object with type information. 

So storing {"ID": 1003.4} and getting {"ID":{"$numberDouble": "1003.4"}} is not expected. Any JSON parser when it tries to unmarshal the second object will fail. 

For a user who manually decodes the returned value into an object this will break ie. return string value of the response instead of object, then manually following MongoDB canonical form, convert returned string into object representation.

**Comment by berndverst** [comment]
I agree with this change.

Currently what we have is:
Save State with Dapr API Data Format but Receive State with Mongo DB API Data format. That is super inconsistent and confusing.

Support for the pure Mongo syntax format belongs into a potential Binding, but definitely not into the State Store.

I support making this "breaking" change. I highly doubt someone would expect to receive the current response output when using the Dapr API.

File: state/mongodb/mongodb.go (hunk 1 above)
Comment (dmitsh) [reply state/mongodb/mongodb.go:228]: > This seems to be a breaking change. If this component is Alpha, it is OK.

This is not a breaking change, but a bug fix.

**Comment by dmitsh** [comment]
This PR should fix querying non-string values: https://github.com/dapr/components-contrib/pull/1584

File: tests/conformance/state/state.go [hunk 2]
```diff
@@ -33,6 +33,10 @@ type ValueType struct {
 	Message string `json:"message"`
 }
 
+type IntValueType struct {
```
Comment (shubham1172) [inline tests/conformance/state/state.go:36]: Can this be unexported?

**Comment by mukundansundar** [comment]
@dapr/maintainers-components-contrib Any call on whether this breaking change can go into existing version of component or not?

File: tests/conformance/state/state.go (hunk 2 above)
Comment (mukundansundar) [reply tests/conformance/state/state.go]: done

**Comment by mukundansundar** [comment]
@dapr/maintainers-components-contrib The conformance test  in PR #1585 as well validates the necessity of  this PR for fixing MongoDB. The call to take is: is it ok if it is a breaking change for few if it is a bug fix .... ?

**Comment by artursouza** [comment]
OK. Let's merge this PR as a fix and mention that in the breaking changes section.

**Pull request merged**